homeassistant/components/insteon/schemas.py

```diff
@@ -187,47 +187,47 @@ def add_device_override(config_data, new_override):
     except ValueError as err:
         raise ValueError("Incorrect values") from err
 
-    overrides = config_data.get(CONF_OVERRIDE, [])
-    curr_override = {}
+    overrides = []
 
-    # If this address has an override defined, remove it
-    for override in overrides:
-        if override[CONF_ADDRESS] == address:
-            curr_override = override
-            break
-    if curr_override:
-        overrides.remove(curr_override)
+    for override in config_data.get(CONF_OVERRIDE, []):
+        if override[CONF_ADDRESS] != address:
+            overrides.append(override)
 
+    curr_override = {}
     curr_override[CONF_ADDRESS] = address
     curr_override[CONF_CAT] = cat
     curr_override[CONF_SUBCAT] = subcat
     overrides.append(curr_override)
-    config_data[CONF_OVERRIDE] = overrides
-    return config_data
+
+    new_config = {}
+    if config_data.get(CONF_X10):
+        new_config[CONF_X10] = config_data[CONF_X10]
+    new_config[CONF_OVERRIDE] = overrides
+    return new_config
 
 
 def add_x10_device(config_data, new_x10):
```

Comment:
Is there an opportunity to break `add_x10_device` and `add_device_override` into smaller units?

When I think about what they do, `add_x10_device` does a lot of things:

1. Iterates the current x10 config and finds only devices that are not the same house/unit code as the new one being added.
2. Creates the new device entry
3. Looks for `CONF_OVERRIDE` and copies it to the new config
4. Merges and returns the new config.

I realize that might not be possible with the current design, but I think its at least worth thinking about.

Reply:
Yeah, you are right. The config entry should be a dict not a list, then I could use the key of the device address to change or add a config entry for that device without the logic of finding the same device in the list. Making this change would also require migrating the existing config to the new format as well, which would be removed at a later date.  But, this would have made a lot of things simpler (doh). 

That would require refactoring and make this PR much bigger. Since this fixes an existing bug I would like to move forward with the code as is and I will commit to putting that in my backlog and getting it done right away.